File: leetcode.com/python/729_My_Calendar_I.py
class TreeNode:
    def __init__(self, start, end):
        self.start = start
        self.end = end
        self.left = None
        self.right = None

    def insert(self, start, end):
        if self.start >= end:
            if self.left:
                return self.left.insert(start, end)
            else:
                self.left = TreeNode(start, end)
                return True
        elif self.end <= start:
            if self.right:
                return self.right.insert(start, end)
            else:
                self.right = TreeNode(start, end)
                return True
        else:
            return False

class MyCalendar(object):
    def __init__(self):
        self.root = None

    def book(self, start, end):
        """
        :type start: int
        :type end: int
        :rtype: bool
        """
        if not self.root:
            self.root = TreeNode(start, end)
            return True
        return self.root.insert(start, end)

# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/729_My_Calendar_I.swift
import Foundation

class TreeNode {
    let start: Int
    let end: Int
    var left: TreeNode?
    var right: TreeNode?
    
    init(_ start: Int, _ end: Int) {
        self.start = start
        self.end = end
    }
}

class MyCalendar {

    var root: TreeNode?
    
    init() {
        
    }
    
    func bookHelper(_ start: Int, _ end: Int, _ node: inout TreeNode) -> Bool {
        if start >= node.end {
            if node.right != nil {
                return self.bookHelper(start, end, &node.right!)
            } else {
                node.right = TreeNode(start, end)
                return true
            }
        } else if end <= node.start {
            if node.left != nil {
                return self.bookHelper(start, end, &node.left!)
            } else {
                node.left = TreeNode(start, end)
                return true
            }
        } else {
            return false
        }
    }
    
    func book(_ start: Int, _ end: Int) -> Bool {
        if let root = self.root {
            return self.bookHelper(start, end, &self.root!)
        }
        self.root = TreeNode(start, end)
        return true
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * let obj = MyCalendar()
 * let ret_1: Bool = obj.book(start, end)
 */
